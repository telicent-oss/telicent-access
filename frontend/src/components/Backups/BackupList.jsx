import { TeliChip, TeliSpinner } from "@telicent-oss/ds";
import React from "react"
import { confirmAlert } from "react-confirm-alert";
import Panel from "../../lib/Panel";

const BackupList = ({ loading, setError, backups, getBackups, deleteSingleBackup, restoreSingleBackup }) => {

  const mostRecentID = backups.reduce((acc, curr) => {
    if (!acc.date || curr.date > acc.date) {
      return curr
    }
    return acc
  }, {})?.id



  const onDeleteHandler = (id) => async () => {
    confirmAlert({
      title: "Delete Backup?",
      message: "This permanently removes the backup.",
      buttons: [
        {
          label: "Delete",
          onClick: async () => {

            // eslint-disable-next-line no-unused-vars
            const [_, err] = await deleteSingleBackup(id)
            if (err) {
              setError(err)
              return
            }
            await getBackups();




          },
        },
        {
          label: "Don't Delete",
        },
      ],
    });
  }



  const onRestoreHandler = (id) => async () => {
    confirmAlert({
      title: `Restore Backup ${id}?`,
      message: "This will cause a temporary outage as we restore to this artefact.",
      buttons: [
        {
          label: "Restore",
          onClick: async () => {

            // eslint-disable-next-line no-unused-vars
            const [_, err] = await restoreSingleBackup(id)
            if (err) {
              setError(err)
            }

          },
        },
        {
          label: "Don't Restore",
        },
      ],
    });
  }
  if (loading) {
    return (
      <div className="mx-12 flex items-center">
        <TeliSpinner size={64} className="flex" /><p className='ml-4'>{loading.message}</p>
      </div>)
  }
  if (backups.length === 0) {
    return (
      <div className="mx-12 flex items-center">
        <p>No backups available</p>
      </div>)
  }
  return (
    <ul className="mx-12">

      {backups.map(backup => (

        <Panel ariaLabel="tile-group" key={backup.id} additionalClassName="!bg-black-100">

          <div className="flex w-full">
            <div className="flex flex-none items-center w-10">
              <div
                className="flex w-10 h-10"

              >
                <i className="self-center w-4 fa-solid fa-folder" />
              </div>
            </div>
            <div className="flex flex-[35%] items-center min-w-0">
              <div className="">
                <div
                  aria-label="text-name"
                  className="truncate font-bold"
                  title={`Backup: ${backup.id}`}
                >
                  Backup: {backup.id}
                </div>
                <div
                  aria-label="text-id"
                  className="truncate text-sm"
                  title={`Date Taken: ${backup.dateTaken}`}
                >
                  {backup.dateTaken}
                </div>
              </div>

              {backup.datasets.map(dataset => <TeliChip
                size="small"
                className="max-w-[150px] ml-6 "
                title={dataset}
                label={dataset}
              />)}
              {mostRecentID === backup.id ?
                <TeliChip
                  size="small"
                  className="max-w-[150px] ml-6 !bg-appColor-hex"
                  title="Most recent backup"
                  label="most recent"
                /> : null}
            </div>

            <div className="flex flex-end items-center">
              <BackupActionButton
                onHandler={onRestoreHandler(backup.id)}
                title="Restore backup"
                label="Restore "
                icon="fa-arrow-up-from-bracket"
                additionalClassName="hover:bg-green-500 pl-1 mr-4"

              />

              <BackupActionButton
                onHandler={onDeleteHandler(backup.id)}
                title="Delete backup"
                icon="fa-trash"
                additionalClassName="hover:bg-red-500  mr-4"

              />

            </div>
          </div>
        </Panel>

      ))
      }
    </ul >
  )
}

const BackupActionButton = ({ onHandler, title, label, icon, additionalClassName }) => {

  const onActionHandler = async () => {


    await onHandler()


  }

  return (
    <button
      type="button"
      aria-label={title}
      title={title}
      className={`flex h-6 rounded ${additionalClassName}
        hover:text-black-100`}
      onClick={onActionHandler}
    >
      {label} <i className={` self-center w-6 fa-solid ${icon}`} />
    </button>
  )
}
export default BackupList;
