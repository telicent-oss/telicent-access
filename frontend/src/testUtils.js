import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

export const renderWithBrowserRouter = (component) => ({
  user: userEvent.setup(),
  ...render(component, { wrapper: BrowserRouter }),
});

export const renderWithUser = (component) => ({
  user: userEvent.setup(),
  ...render(component, { wrapper: BrowserRouter }),
});

export const getLabels = (user) => {
  const labels = {};
  user.labels.forEach(({ name, value }) => {
    labels[name] = value;
  });
  return labels;
};

export const countries = [
  { name: "Japan", alpha2: "JP", alpha3: "JPN" },
  { name: "Sweden", alpha2: "SE", alpha3: "SWE" },
  { name: "United Kingdom", alpha2: "GB", alpha3: "GBR" },
];

export const exampleUsers = [
  {
    id: "64e3690cbcd5a5eb87b82836",
    externalId: "8bb01245-261e-4609-a4a5-xxxxxxxxxxxx",
    name: "Admin User",
    userName: "admin@telicent.io",
    email: "admin@telicent.io",
    labels: [
      {
        name: "nationality",
        value: "GBR",
        toString: 'nationality="GBR"',
        toDataLabelString: "permitted_nationalities='GBR'",
        _id: "64e36950bcd5a5eb87b8284b",
      },
      {
        name: "clearance",
        value: "S",
        toString: 'clearance="S"',
        toDataLabelString: "classification='S'",
        _id: "64e36950bcd5a5eb87b8284c",
      },
      {
        name: "personnel_type",
        value: "NON-GOV",
        toString: 'personnel_type="NON-GOV"',
        toDataLabelString: null,
        _id: "64e36950bcd5a5eb87b8284d",
      },
      {
        name: "deployed_organisation",
        value: "Telicent",
        toString: 'deployed_organisation="Telicent"',
        toDataLabelString: "permitted_organisations='Telicent'",
        _id: "64e36950bcd5a5eb87b8284e",
      },
    ],
    active: true,
    groups: [],
    userGroups: ["urn:telicent:groups:c-suite"],
    schemas: ["urn:ietf:params:scim:schemas:core:2.0:User"],
  },
  {
    id: "64e3690cbcd5a5eb87b82835",
    externalId: "138c92a8-cbc8-4a4c-822c-zzzzzzzzzzzz",
    name: "Test User",
    userName: "test@telicent.io",
    email: "test@telicent.io",
    labels: [
      {
        name: "nationality",
        value: "GBR",
        toString: 'nationality="GBR"',
        toDataLabelString: "permitted_nationalities='GBR'",
        _id: "64e36950bcd5a5eb87b8284b",
      },
      {
        name: "clearance",
        value: "O",
        toString: 'clearance="O"',
        toDataLabelString: "classification='O'",
        _id: "64e36950bcd5a5eb87b8284c",
      },
      {
        name: "personnel_type",
        value: "NON-GOV",
        toString: 'personnel_type="NON-GOV"',
        toDataLabelString: null,
        _id: "64e36950bcd5a5eb87b8284d",
      },
      {
        name: "deployed_organisation",
        value: "Telicent",
        toString: 'deployed_organisation="Telicent"',
        toDataLabelString: "permitted_organisations='Telicent'",
        _id: "64e36950bcd5a5eb87b8284e",
      },
    ],
    active: true,
    groups: [],
    userGroups: ["urn:telicent:groups:c-suite"],
    schemas: ["urn:ietf:params:scim:schemas:core:2.0:User"],
  },
];

export const exampleUser = exampleUsers[0];

export const exampleLabels = [
  {
    _id: "clearance-test-id",
    name: "clearance",
    usage: "both",
    value: {
      type: "hierarchy",
      values: ["O", "OS", "S", "TS"],
    },
    ihm: true,
    readonly: true,
    user_required: true,
  },
  {
    _id: "nationality-test-id",
    name: "nationality",
    usage: "both",
    value: {
      type: "enum",
      values: countries.map(({ alpha3 }) => alpha3.trim()),
    },
    ihm: true,
    readonly: true,
    user_required: true,
  },
  {
    _id: "org-test-id",
    name: "deployed_organisation",
    usage: "both",
    value: {
      type: "string",
      values: null,
    },
    ihm: true,
    readonly: true,
    user_required: true,
  },
  {
    _id: "personnel-type-test-id",
    name: "personnel_type",
    usage: "both",
    value: {
      type: "enum",
      values: ["GOV", "NON-GOV"],
    },
    ihm: true,
    readonly: true,
    user_required: true,
  },
];

export const exampleAttrs = [
  {
    value: { type: "hierarchy", values: ["O", "OS", "S", "TS"] },
    _id: "clearance-test-id",
    user_attribute_name: "clearance",
    data_attribute_name: "classification",
    ihm: true,
    readonly: true,
  },
  {
    value: {
      type: "enum",
      values: ["AFG", "ALB", "DZA", "GBR"],
    },
    _id: "nationality-test-id",
    user_attribute_name: "nationality",
    data_attribute_name: "permitted_nationalities",
    ihm: true,
    readonly: true,
  },
  {
    value: { type: "string", values: null },
    _id: "org-test-id",
    user_attribute_name: "deployed_organisation",
    data_attribute_name: "permitted_organisations",
    ihm: true,
    readonly: true,
  },
  {
    value: { type: "enum", values: ["GOV", "NON-GOV"] },
    _id: "personnel-type-test-id",
    user_attribute_name: "personnel_type",
    data_attribute_name: null,
    ihm: true,
    readonly: true,
  },
];

export const exampleGroups = [
  {
    _id: "64d2783ec248737ae094c4e8",
    label: "surgical",
    group_id: "urn:telicent:groups:surgical",
    description: "Group for all managers",
    active: true,
  },
  {
    _id: "64d27846c248737ae094c4ea",
    label: "manager",
    group_id: "urn:telicent:groups:manager",
    description: "Group for all managers",
    active: false,
  },
  {
    _id: "64d27aae0e121de092c20621",
    label: "developers",
    group_id: "urn:telicent:groups:developers",
    description: "Group for all developers",
    active: true,
  },
  {
    _id: "64d27b1c588b7f231aa924ef",
    label: "c-suite",
    group_id: "urn:telicent:groups:c-suite",
    description: "Group for all c-suite",
    active: true,
  },
];

export const exampleHierarchy = {
  uuid: "31127653-a234-48bc-b940-34c866bfe384",
  name: "clearance",
  tiers: ["O", "S", "TS"],
  readonly: true,
};
