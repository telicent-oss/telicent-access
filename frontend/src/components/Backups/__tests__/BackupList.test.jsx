import { render, screen, waitFor } from "@testing-library/react"
import { act } from "react";
import BackupList from "../BackupList"
import { renderWithBrowserRouter } from "../../../testUtils";

describe("BackupList", () => {

  it("should show no backups available", () => {
    render(<BackupList backups={[]} loading={false} setError={jest.fn()} deleteSingleBackup={jest.fn()} restoreSingleBackup={jest.fn()} getBackups={jest.fn()} />);

    expect(screen.getByText("No backups available")).toBeInTheDocument();
  });

  it("should show list available backups", () => {
    const sampleBackups = [
      {
        id: "1",
        dateTaken: "2025-06-24_13-57-46",
        datasets: ["catalog"]
      }
    ]

    const { container } = render(<BackupList backups={sampleBackups} loading={false} setError={jest.fn()} deleteSingleBackup={jest.fn()} restoreSingleBackup={jest.fn()} getBackups={jest.fn()} />);

    expect(container).toMatchSnapshot()
  });

  it.skip("should pass backup id to restore fn", async () => {
    const sampleBackups = [
      {
        id: "1",
        dateTaken: "2025-06-24_13-57-46",
        datasets: ["catalog"]
      }
    ]

    const restoreFnMock = jest.fn();
    const { user, container } = renderWithBrowserRouter(<BackupList backups={sampleBackups} loading={false} setError={jest.fn()} deleteSingleBackup={jest.fn()} restoreSingleBackup={restoreFnMock} getBackups={jest.fn()} />);

    const restoreButton = container.querySelector("button[title=\"Restore Backup\"]")
    await user.click(restoreButton);

    const confirmRestoreButton = await screen.findByText("Restore");

    await act(async () => {
      await user.click(confirmRestoreButton);
    })

    expect(restoreFnMock).toHaveBeenCalledWith("1");
  });

});
