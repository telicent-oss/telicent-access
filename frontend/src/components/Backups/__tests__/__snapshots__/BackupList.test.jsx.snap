// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackupList should show list available backups 1`] = `
<div>
  <ul
    class="mx-12"
  >
    <li
      aria-label="tile-group"
      class="flex h-fit w-full mb-3 py-4 px-6 bg-black-200 rounded-lg !bg-black-100"
    >
      <div
        class="flex w-full"
      >
        <div
          class="flex flex-none items-center w-10"
        >
          <div
            class="flex w-10 h-10"
          >
            <i
              class="self-center w-4 fa-solid fa-folder"
            />
          </div>
        </div>
        <div
          class="flex flex-[35%] items-center min-w-0"
        >
          <div
            class=""
          >
            <div
              aria-label="text-name"
              class="truncate font-bold"
              title="Backup: 1"
            >
              Backup: 
              1
            </div>
            <div
              aria-label="text-id"
              class="truncate text-sm"
              title="Date Taken: 2025-06-24_13-57-46"
            >
              2025-06-24_13-57-46
            </div>
          </div>
          <div
            class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault teli-chip--primary max-w-[150px] ml-6 css-13htn8f-MuiChip-root"
            title="catalog"
          >
            <span
              class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
            >
              catalog
            </span>
          </div>
          <div
            class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault teli-chip--primary max-w-[150px] ml-6 !bg-appColor-hex css-13htn8f-MuiChip-root"
            title="Most recent backup"
          >
            <span
              class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
            >
              most recent
            </span>
          </div>
        </div>
        <div
          class="flex flex-end items-center"
        >
          <button
            aria-label="Restore backup"
            class="flex h-6 rounded hover:bg-green-500 pl-1 mr-4
        hover:text-black-100"
            title="Restore backup"
            type="button"
          >
            Restore 
             
            <i
              class=" self-center w-6 fa-solid fa-arrow-up-from-bracket"
            />
          </button>
          <button
            aria-label="Delete backup"
            class="flex h-6 rounded hover:bg-red-500  mr-4
        hover:text-black-100"
            title="Delete backup"
            type="button"
          >
             
            <i
              class=" self-center w-6 fa-solid fa-trash"
            />
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
